File: IT/Oggetti/Java/CopyCtor.java
/* file CopyCtor.java */

import java.util.Map;
import java.util.Map.Entry;
import java.util.HashMap;

public class CopyCtor {
	private static class Student {
		private String name;
		private String matr;
		private String cdl;
		private HashMap<String, Integer> es;

		public Student(String name, String matr, String cdl) {
			this.name = name;
			this.matr = matr;
			this.cdl = cdl;
			this.es = new HashMap<String, Integer>() {{
				put("Architettura degli elaboratori", -1);
				put("Programmazione", -1);
				put("Calcolo 1", -1);
			}};
		}

		// nuovo costruttore di copia

		public Student(Student s) {
			this.name = s.name;
			this.matr = s.matr;
			this.cdl = s.cdl;
			this.es = new HashMap<String, Integer>(s.es);
		}

		public void unilib() {
			int _pass = 0;

			for (Map.Entry<String, Integer> p : this.es.entrySet()) {
				String k = p.getKey();
				Integer v = p.getValue(); 
				if (v != -1) {
					System.out.println(this.name + " (" + this.matr + ") ha superato l'esame \"" + k + "\" con il voto " + v + ".");
				_pass++;
				}
			}
		}

		@Override
		public String toString() {
			return this.name + ", matricola " + this.matr + ", corso di laurea in " + this.cdl + ".";
		}
	}

	public static void main(String [] args) {
		Student s1 = new Student("Mario Rossi", "A23234", "Scienze dell'Informazione");
		System.out.println(s1);

		// crea un nuovo studente con il costruttore di copia

		Student s2 = new Student(s1);

		s2.matr = "A01357";
		s2.cdl = "Ingegneria Informatica";
		s2.es.put("Programmazione", 30);

		System.out.println(s2);
		s1.unilib();
		s2.unilib();
	}
}

File: IT/1/Oggetti/C++/copyCtor.cpp
/*
file copyCtor.cpp
compilare con: g++ -std=c++11 testCopyCtor.cpp
*/

#include <iostream>
#include <ostream>
#include <unordered_map>

using std::ostream;
using std::unordered_map;
using std::string;
using std::cout;
using std::endl;

class Student {
	public:
		string name;
		string matr;
		string cdl;
		unordered_map<string, int> es;

	public:
		Student() = default;
		Student(const char*, const char*, const char*);
		// costruttore di copia
		Student(const Student& s);
		void unilib();

	Student& operator=(const Student&);
	friend std::ostream& operator<<(std::ostream&, const Student&);
};

Student::Student(const char* name, const char* matr, const char* cdl) : es {{"Architettura degli elaboratori", -1},  {"Programmazione", -1}, {"Calcolo 1", -1}} {
	this->name = name;
	this->matr = matr;
	this->cdl = cdl;
}

Student::Student(const Student& s) {
	this->name = s.name;
	this->matr = s.matr;
	this->cdl = s.cdl;
	this->es = s.es;
}

Student& Student::operator=(const Student& s) {
	if (this == &s) return *this;
	this->name = s.name;
	this->matr = s.matr;
	this->cdl = s.cdl;
	this->es = s.es;
	return *this;
}

ostream& operator<<(ostream &os, const Student& s) { 
	os << s.name << ", matricola " << s.matr << ", corso di laurea in " << s.cdl << "."; 
	return os;
}

void Student::unilib() {
	int _pass = 0;
	for (auto& p : this->es) {
		string k = p.first;
		int v = p.second;
		if (v != -1) {
			cout << this->name << " (" << this->matr << ") ha superato l'esame \"" << k << "\" con il voto " << v << "." << endl;
			_pass++;
		}
	}
}

int main() {
	Student s1("Mario Rossi", "A23234", "Scienze dell'Informazione");

	// crea un nuovo studente con il costruttore di copia
	Student s2 = s1;

	// in alternativa si può usare l'operatore di copia e assegnamento
	// Student s2;
	// s2 = s1;

	s2.matr = "A01357";
	s2.cdl = "Ingegneria Informatica";
	s2.es["Programmazione"] = 30;
	cout << s1 << endl;
	cout << s2 << endl;
	s1.unilib();
	s2.unilib();
}

